Catch started but not finished tasks

<!-- 
Thank you for your interest in this project. Bugs filed and PRs submitted are appreciated!

Please make sure that you are familiar with and follow the Code of Conduct of this project which can be found at https://github.com/facebookincubator/TestSlide/blob/master/CODE_OF_CONDUCT.md 

Also, please make sure you're familiar with and follow the instructions in the contributing guidelines which can  be found at https://github.com/facebookincubator/TestSlide/blob/master/CONTRIBUTING.md

- Please review the Issues policies before filing an issue:

  🐛 Bug:
  Report a bug, missing documentation, or unexpected behavior.

-->

**I'm using `TestSlide` version:** 2.5.9

**Given:** Code that creates an asyncio task but does not wait for it (see below):
<!-- 
Include code snippets/a link to online editor like CodeSandbox/configurations/anything
-->
```python
#!/usr/bin/env python3
import asyncio
import sys
from testslide.dsl import context


def to_mock():
    pass


async def some_async_function():
    to_mock()


@context
def async_test(context):
    @context.example
    async def test_assert_called_fails(self):
        self.mock_callable(sys.modules[__name__], 'to_mock')\
            .to_return_value(None)\
            .and_assert_called_once()
        asyncio.create_task(some_async_function())

    @context.example
    async def test_assert_called_works_when_extra_await(self):
        self.mock_callable(sys.modules[__name__], 'to_mock')\
            .to_return_value(None)\
            .and_assert_called_once()
        asyncio.create_task(some_async_function())
        await asyncio.sleep(1)
```

**When I run:** testslide example.py
<!-- 
Include run/execution commands
-->
```
 # testslide example.py 
async test
  test assert called fails: AssertionError: calls did not match assertion.
  test assert called works when extra await

Failures:

  1) async test: test assert called fails
    1) AssertionError: calls did not match assertion.
    <module 'example' from '/home/dcaro/example.py'>, 'to_mock':
      expected: called exactly 1 time(s) with any arguments   received: 0 call(s)
      File "work/TestSlide/testslide/mock_callable.py", line 250, in assertion
        raise AssertionError(

Finished 2 example(s) in 2.1s 
  Successful: 1
  Failed: 1
```

**I expected this to happen:** Testslide should tell me that there were non-finished tasks when the test finished, maybe just a warning.

<!-- Describe the correct/expected behavior -->

**But, instead this happened:** Testslide just fails as if the function was not called (as it wasn't, because the task ran/was cancelled after the test finished).

<!-- 
Describe the unexpected behavior that has happened. Include complete error messages/screenshots/anything
-->

--- a/docs/testslide_dsl/async_support/index.rst
+++ b/docs/testslide_dsl/async_support/index.rst
@@ -43,7 +43,7 @@ like this:
       assert self.memoize_before == "memoize_before"
       assert self.function == "function"
 
-The test runner will create a new event look to execute each example.
+The test runner will create a new event loop to execute each example.
 
 .. note::
 
@@ -184,3 +184,36 @@ Python's default threshold for triggering this event loop lock up failure is **1
 
   Finished 1 example(s) in 1.0s
     Successful: 1
+
+Leaked Tasks
+^^^^^^^^^^^^
+
+If your async code creates a task in the asyncio loop, but finished before that task has ended (ex. you forgot to await for it), testslide will catch it and fail the test.
+
+This is enabled by default for async tests, but to get that behaviour also when running async code from sync tests, for example:
+
+
+.. code-block:: python
+
+  import asyncio
+  from testslide.dsl import context
+
+  @context
+  def my_test_suite(context):
+        @context.example
+        def test_something_async(self):
+            asyncio.run(my_async_function())
+
+
+Has to use the `async_run` function from the context, so instead, you should use:
+
+.. code-block:: python
+
+  import asyncio
+  from testslide.dsl import context
+
+  @context
+  def my_test_suite(context):
+        @context.example
+        def test_something_async(self):
+            self.async_run(my_async_function())

--- a/docs/testslide_dsl/async_support/index.rst
+++ b/docs/testslide_dsl/async_support/index.rst
@@ -205,7 +205,7 @@ This is enabled by default for async tests, but to get that behaviour also when 
             asyncio.run(my_async_function())
 
 
-Has to use the `async_run` function from the context, so instead, you should use:
+Has to use the `async_run_with_health_checks` function from the context, so instead, you should use:
 
 .. code-block:: python
 
@@ -216,4 +216,4 @@ Has to use the `async_run` function from the context, so instead, you should use
   def my_test_suite(context):
         @context.example
         def test_something_async(self):
-            self.async_run(my_async_function())
+            self.async_run_with_health_checks(my_async_function())

--- a/tests/dsl_unittest.py
+++ b/tests/dsl_unittest.py
@@ -2073,13 +2073,63 @@ class TestMockConstructorIntegration(TestDSLBase):
                     ).to_return_value("mocked_popen").and_assert_called_once()
                     assert subprocess.Popen(["cmd"]) == "mocked_popen"
 
-        examples = {}
-
-        for all_top_level_context in Context.all_top_level_contexts:
-            for example in all_top_level_context.all_examples:
-                examples[example.name] = example
+        examples = _get_name_to_examples()
 
         self.run_example(examples["expect pass"])
 
         with self.assertRaisesRegex(AssertionError, "calls did not match assertion"):
             self.run_example(examples["expect fail"])
+
+
+class TestMock(TestDSLBase):
+    def test_mock_constructor_integration(self):
+        @context
+        def fail_top(context):
+            @context.sub_context
+            def fail_sub_context(context):
+                @context.example
+                def expect_fail(self):
+                    self.mock_constructor("subprocess", "Popen").for_call(
+                        ["cmd"]
+                    ).to_return_value("mocked_popen").and_assert_called_once()
+
+        @context
+        def pass_top(context):
+            @context.sub_context
+            def pass_sub_context(context):
+                @context.example
+                def expect_pass(self):
+                    self.mock_constructor("subprocess", "Popen").for_call(
+                        ["cmd"]
+                    ).to_return_value("mocked_popen").and_assert_called_once()
+                    assert subprocess.Popen(["cmd"]) == "mocked_popen"
+
+        examples = _get_name_to_examples()
+
+        self.run_example(examples["expect pass"])
+
+        with self.assertRaisesRegex(AssertionError, "calls did not match assertion"):
+            self.run_example(examples["expect fail"])
+
+
+class TestAsyncRun(TestDSLBase):
+    def test_catches_leaked_tasks(self):
+        async def dummy_async_func():
+            pass
+
+        async def spawn_task_but_dont_await():
+            # once we stop supporting python 3.6 we can use
+            # asyncio.create_task instead
+            loop = asyncio.get_event_loop()
+            loop.create_task(dummy_async_func())
+
+        @context
+        def fail_top(context):
+            @context.example
+            def raise_on_leaked_task(self):
+                self.async_run(spawn_task_but_dont_await())
+
+        examples = _get_name_to_examples()
+
+        with self.assertRaisesRegex(RuntimeError, "Some tasks were started"):
+            self.run_example(examples["raise on leaked task"])

--- a/tests/dsl_unittest.py
+++ b/tests/dsl_unittest.py
@@ -2040,10 +2040,11 @@ class TestMockAsyncCallableIntegration(TestDSLBase):
         def fail_top(context):
             @context.example
             async def spawn_task_but_dont_await(self):
-                # once we stop supporting python 3.6 we can use
-                # asyncio.create_task instead
-                loop = asyncio.get_event_loop()
-                loop.create_task(dummy_async_func())
+                if sys.version_info < (3, 7):
+                    loop = asyncio.get_event_loop()
+                    loop.create_task(dummy_async_func())
+                else:
+                    asyncio.create_task(dummy_async_func())
 
         examples = _get_name_to_examples()
 
@@ -2088,10 +2089,11 @@ class TestAsyncRun(TestDSLBase):
             pass
 
         async def spawn_task_but_dont_await():
-            # once we stop supporting python 3.6 we can use
-            # asyncio.create_task instead
-            loop = asyncio.get_event_loop()
-            loop.create_task(dummy_async_func())
+            if sys.version_info < (3, 7):
+                loop = asyncio.get_event_loop()
+                loop.create_task(dummy_async_func())
+            else:
+                asyncio.create_task(dummy_async_func())
 
         @context
         def fail_top(context):

--- a/tests/dsl_unittest.py
+++ b/tests/dsl_unittest.py
@@ -1621,7 +1621,7 @@ class TestExample(TestDSLBase):
 
     def test_inherits_skip_from_xcontext(self):
         """
-        Exmaples inherit skip setting from parent context.
+        Examples inherit skip setting from parent context.
         """
 
         @xcontext
@@ -1647,7 +1647,7 @@ class TestExample(TestDSLBase):
 
     def test_inherits_focus_from_fcontext(self):
         """
-        Exmaples inherit focus setting from parent context.
+        Examples inherit focus setting from parent context.
         """
 
         @fcontext

--- a/tests/dsl_unittest.py
+++ b/tests/dsl_unittest.py
@@ -2081,37 +2081,6 @@ class TestMockConstructorIntegration(TestDSLBase):
             self.run_example(examples["expect fail"])
 
 
-class TestMock(TestDSLBase):
-    def test_mock_constructor_integration(self):
-        @context
-        def fail_top(context):
-            @context.sub_context
-            def fail_sub_context(context):
-                @context.example
-                def expect_fail(self):
-                    self.mock_constructor("subprocess", "Popen").for_call(
-                        ["cmd"]
-                    ).to_return_value("mocked_popen").and_assert_called_once()
-
-        @context
-        def pass_top(context):
-            @context.sub_context
-            def pass_sub_context(context):
-                @context.example
-                def expect_pass(self):
-                    self.mock_constructor("subprocess", "Popen").for_call(
-                        ["cmd"]
-                    ).to_return_value("mocked_popen").and_assert_called_once()
-                    assert subprocess.Popen(["cmd"]) == "mocked_popen"
-
-        examples = _get_name_to_examples()
-
-        self.run_example(examples["expect pass"])
-
-        with self.assertRaisesRegex(AssertionError, "calls did not match assertion"):
-            self.run_example(examples["expect fail"])
-
-
 class TestAsyncRun(TestDSLBase):
     def test_catches_leaked_tasks(self):
         async def dummy_async_func():

--- a/tests/dsl_unittest.py
+++ b/tests/dsl_unittest.py
@@ -24,6 +24,15 @@ from testslide.dsl import context, fcontext, xcontext
 from testslide.runner import QuietFormatter
 
 
+def _get_name_to_examples():
+    examples = {}
+    for all_top_level_context in Context.all_top_level_contexts:
+        for example in all_top_level_context.all_examples:
+            examples[example.name] = example
+
+    return examples
+
+
 class SomeTestCase(unittest.TestCase):
     """
     Used to test TestSlide and unittest.TestCase integration.
@@ -1950,11 +1959,7 @@ class TestPatchAttributeIntegration(TestDSLBase):
             def unpatching_works(self):
                 self.assertEqual(SomeClass.attribute, "original_value")
 
-        examples = {}
-
-        for all_top_level_context in Context.all_top_level_contexts:
-            for example in all_top_level_context.all_examples:
-                examples[example.name] = example
+        examples = _get_name_to_examples()
 
         self.run_example(examples["can patch attribute"])
         self.run_example(examples["unpatching works"])
@@ -1983,11 +1988,7 @@ class TestMockCallableIntegration(TestDSLBase):
                     ).and_assert_called_once()
                     assert os.getcwd() == "mocked_cwd"
 
-        examples = {}
-
-        for all_top_level_context in Context.all_top_level_contexts:
-            for example in all_top_level_context.all_examples:
-                examples[example.name] = example
+        examples = _get_name_to_examples()
 
         self.run_example(examples["expect pass"])
 
@@ -2023,16 +2024,30 @@ class TestMockAsyncCallableIntegration(TestDSLBase):
                     ).for_call().to_return_value("mocked").and_assert_called_once()
                     assert await SomeClass.do_something() == "mocked"
 
-        examples = {}
-
-        for all_top_level_context in Context.all_top_level_contexts:
-            for example in all_top_level_context.all_examples:
-                examples[example.name] = example
+        examples = _get_name_to_examples()
 
         self.run_example(examples["expect pass"])
 
         with self.assertRaisesRegex(AssertionError, "calls did not match assertion"):
             self.run_example(examples["expect fail"])
+
+    def test_leaked_tasks_fails_test_run(self):
+        async def dummy_async_func():
+            pass
+
+        @context
+        def fail_top(context):
+            @context.example
+            async def spawn_task_but_dont_await(self):
+                # once we stop supporting python 3.6 we can use
+                # asyncio.create_task instead
+                loop = asyncio.get_event_loop()
+                loop.create_task(dummy_async_func())
+
+        examples = _get_name_to_examples()
+
+        with self.assertRaisesRegex(RuntimeError, "Some tasks were started"):
+            self.run_example(examples["spawn task but dont await"])
 
 
 class TestMockConstructorIntegration(TestDSLBase):

--- a/tests/dsl_unittest.py
+++ b/tests/dsl_unittest.py
@@ -16,6 +16,7 @@ from testslide import (
     AggregatedExceptions,
     Context,
     Example,
+    LeftOverActiveTasks,
     SlowCallback,
     _ExampleRunner,
     reset,
@@ -2046,7 +2047,7 @@ class TestMockAsyncCallableIntegration(TestDSLBase):
 
         examples = _get_name_to_examples()
 
-        with self.assertRaisesRegex(RuntimeError, "Some tasks were started"):
+        with self.assertRaisesRegex(LeftOverActiveTasks, "Some tasks were started"):
             self.run_example(examples["spawn task but dont await"])
 
 
@@ -2100,5 +2101,5 @@ class TestAsyncRun(TestDSLBase):
 
         examples = _get_name_to_examples()
 
-        with self.assertRaisesRegex(RuntimeError, "Some tasks were started"):
+        with self.assertRaisesRegex(LeftOverActiveTasks, "Some tasks were started"):
             self.run_example(examples["raise on leaked task"])

--- a/tests/dsl_unittest.py
+++ b/tests/dsl_unittest.py
@@ -2127,7 +2127,7 @@ class TestAsyncRun(TestDSLBase):
         def fail_top(context):
             @context.example
             def raise_on_leaked_task(self):
-                self.async_run(spawn_task_but_dont_await())
+                self.async_run_with_health_checks(spawn_task_but_dont_await())
 
         examples = _get_name_to_examples()
 

--- a/testslide/__init__.py
+++ b/testslide/__init__.py
@@ -20,6 +20,7 @@ import types
 import unittest
 import warnings
 from contextlib import contextmanager
+from functools import partial
 from typing import (
     TYPE_CHECKING,
     Any,
@@ -46,6 +47,24 @@ if TYPE_CHECKING:
 
 if sys.version_info < (3, 6):
     raise RuntimeError("Python >=3.6 required.")
+
+if sys.version_info < (3, 7):
+
+    def asyncio_run(coro):
+        loop = asyncio.events.new_event_loop()
+        try:
+            loop.set_debug(True)
+            loop.run_until_complete(coro)
+        finally:
+            try:
+                loop.run_until_complete(loop.shutdown_asyncgens())
+            finally:
+                loop.close()
+
+
+else:
+    asyncio_run = partial(asyncio.run, debug=True)
+
 
 if sys.version_info < (3, 8):
     get_all_tasks = asyncio.Task.all_tasks
@@ -210,6 +229,15 @@ class _ContextData(object):
         """
         with self._sub_examples_agg_ex.catch():
             yield
+
+    def async_run(self, coro):
+        """
+        Runs the given coroutine in a new event loop, and ensuring there's no
+        task leakage.
+        """
+        result = asyncio_run(_async_ensure_no_leaked_tasks(coro))
+
+        return result
 
 
 class AggregatedExceptions(Exception):
@@ -422,18 +450,7 @@ class _ExampleRunner:
     def _async_run_all_hooks_and_example(self, context_data: _ContextData) -> None:
         coro = self._real_async_run_all_hooks_and_example(context_data)
         with self._raise_if_asyncio_warnings(context_data):
-            if sys.version_info < (3, 7):
-                loop = asyncio.events.new_event_loop()
-                try:
-                    loop.set_debug(True)
-                    loop.run_until_complete(coro)
-                finally:
-                    try:
-                        loop.run_until_complete(loop.shutdown_asyncgens())
-                    finally:
-                        loop.close()
-            else:
-                asyncio.run(coro, debug=True)
+            asyncio_run(coro)
 
     @staticmethod
     def _fail_if_coroutine_function(

--- a/testslide/__init__.py
+++ b/testslide/__init__.py
@@ -230,7 +230,7 @@ class _ContextData(object):
         with self._sub_examples_agg_ex.catch():
             yield
 
-    def async_run(self, coro):
+    def async_run_with_health_checks(self, coro):
         """
         Runs the given coroutine in a new event loop, and ensuring there's no
         task leakage.

--- a/testslide/__init__.py
+++ b/testslide/__init__.py
@@ -47,6 +47,11 @@ if TYPE_CHECKING:
 if sys.version_info < (3, 6):
     raise RuntimeError("Python >=3.6 required.")
 
+if sys.version_info < (3, 8):
+    get_all_tasks = asyncio.Task.all_tasks
+else:
+    get_all_tasks = asyncio.all_tasks
+
 
 def _importer(target: str) -> Any:
     components = target.split(".")
@@ -64,6 +69,21 @@ def _importer(target: str) -> Any:
         import_path += ".%s" % comp
         thing = dot_lookup(thing, comp, import_path)
     return thing
+
+
+async def _async_ensure_no_leaked_tasks(coro):
+    before_example_tasks = get_all_tasks()
+    result = await coro
+    after_example_tasks = get_all_tasks()
+    new_still_running_tasks = set(after_example_tasks) - set(before_example_tasks)
+    if new_still_running_tasks:
+        tasks_str = "\n".join(str(task) for task in new_still_running_tasks)
+        raise RuntimeError(
+            "Some tasks were started but did not finish yet, are you missing "
+            f"an `await` somewhere?\nRunning tasks:\n {tasks_str}"
+        )
+
+    return result
 
 
 class _ContextData(object):
@@ -303,21 +323,24 @@ class _ExampleRunner:
                         before_code, context_data
                     )
                 self.formatter.dsl_example(self.example, self.example.code)
-                await self._fail_if_not_coroutine_function(
-                    self.example.code, context_data
+                await _async_ensure_no_leaked_tasks(
+                    self._fail_if_not_coroutine_function(
+                        self.example.code, context_data
+                    )
                 )
             after_functions: List[Callable] = []
             after_functions.extend(context_data._mock_callable_after_functions)
             after_functions.extend(self.example.context.all_after_functions)
             after_functions.extend(context_data._after_functions)
+
             for after_code in reversed(after_functions):
                 with aggregated_exceptions.catch():
                     self.formatter.dsl_after(self.example, after_code)
                     await self._fail_if_not_coroutine_function(after_code, context_data)
             aggregated_exceptions.raise_correct_exception()
             return
-        around_code = around_functions.pop()
 
+        around_code = around_functions.pop()
         wrapped_called: List[bool] = []
 
         async def async_wrapped() -> None:

--- a/testslide/__init__.py
+++ b/testslide/__init__.py
@@ -72,6 +72,12 @@ else:
     get_all_tasks = asyncio.all_tasks
 
 
+class LeftOverActiveTasks(BaseException):
+    """Risen when unfinished asynchronous tasks are detected."""
+
+    pass
+
+
 def _importer(target: str) -> Any:
     components = target.split(".")
     import_path = components.pop(0)
@@ -97,7 +103,7 @@ async def _async_ensure_no_leaked_tasks(coro):
     new_still_running_tasks = set(after_example_tasks) - set(before_example_tasks)
     if new_still_running_tasks:
         tasks_str = "\n".join(str(task) for task in new_still_running_tasks)
-        raise RuntimeError(
+        raise LeftOverActiveTasks(
             "Some tasks were started but did not finish yet, are you missing "
             f"an `await` somewhere?\nRunning tasks:\n {tasks_str}"
         )

